crucible-syntax/src/Lang/Crucible/Syntax/Monad.hs

```diff
@@ -0,0 +1,524 @@
+  , sideCondition'
+  , satisfy
+
+  -- * Eliminating location information
+  , syntaxToDatum
+  , datum
+
+  -- * Parsing context
+  , position
+  , withProgressStep
+
+  -- * Control structures
+  , commit
+  , parse
+
+  -- * Progress through a parsing problem
+  , ProgressStep(..)
+  , Progress
+  , emptyProgress
+  , pushProgress
+
+  -- * Errors
+  , later
+  , Reason(..)
+  ) where
+
+import Control.Applicative
+import Control.Monad (MonadPlus(..), ap)
+import Control.Monad.Reader (MonadReader(..), ReaderT(..))
+import qualified Control.Monad.State.Strict as Strict
+import qualified Control.Monad.State.Lazy as Lazy
+import Control.Monad.State.Class
+import Control.Monad.Trans.Class (MonadTrans(..))
+import qualified Control.Monad.Writer.Strict as Strict
+import qualified Control.Monad.Writer.Lazy as Lazy
+import Control.Monad.Writer.Class
+
+import Data.Foldable as Foldable
+import Data.List.NonEmpty (NonEmpty(..))
+import Data.Text (Text)
+import qualified Data.Text as T
+
+import Lang.Crucible.Syntax.SExpr
+
+import What4.ProgramLoc (Posd(..), Position)
+
+data Search a = Try a (Search a) | Fail | Cut
+  deriving Functor
+
+instance Applicative Search where
+  pure x = Try x Fail
+  (<*>) = ap
+
+instance Alternative Search where
+  empty = Fail
+  x <|> y =
+    case x of
+      Try first rest -> Try first (rest <|> y)
+      Fail -> y
+      Cut -> Cut
+
+instance Monad Search where
+  m >>= f =
+    case m of
+      Try x more -> f x <|> (more >>= f)
+      Fail -> Fail
+      Cut -> Fail
+
+instance MonadPlus Search where
+  mzero = empty
+  mplus = (<|>)
+
+instance Semigroup (Search a) where
+  (<>) = (<|>)
+
+instance Monoid (Search a) where
+  mempty  = empty
+
+instance Foldable Search where
+  foldMap f (Try x xs) = f x `mappend` foldMap f xs
+  foldMap _ _ = mempty
+
+  toList (Try x xs) = x : toList xs
+  toList _          = []
+
+instance Traversable Search where
+  traverse f (Try x xs) = Try <$> f x <*> traverse f xs
+  traverse _ Fail = pure Fail
+  traverse _ Cut = pure Cut
+
+-- | Components of a path taken through a syntax object to reach the
+-- current focus.
+data ProgressStep =
+    First -- ^ The head of a list was followed
+  | Rest -- ^ The tail of a list was followed
+  | Late -- ^ The path was annotated as 'later'
+  deriving (Eq, Show)
+
+instance Ord ProgressStep where
+  compare First First = EQ
+  compare First _ = LT
+  compare Rest First = GT
+  compare Rest Rest = EQ
+  compare Rest _ = LT
+  compare Late Late = EQ
+  compare Late _ = GT
+
+-- | The path taken through a syntax object to reach the current
+-- focus.
+newtype Progress = Progress [ProgressStep]
+  deriving (Eq, Show)
+
+-- | An empty path, used to initialize parsers
+emptyProgress :: Progress
+emptyProgress = Progress []
+
+-- | Add a step to a progress path
+pushProgress :: ProgressStep -> Progress -> Progress
+pushProgress p (Progress ps) = Progress (p : ps)
+
+instance Ord Progress where
+  compare (Progress xs) (Progress ys) =
```

Comment:
This might be  `compare (Progress xs) (Progress ys) = compare xs ys` except that it first scans lengths so (by possibly incorrect inspection) it yields `Progress [Late, Late] < Progress [First, Rest, Rest]`  which is the opposite of what I would expect from `compare xs ys`.  A Haddock comment about the expected `Ord` behavior would help, even if my inspection analysis here is wrong.